### PR TITLE
feat: add `Metadata.findDataByOwner`

### PR DESCRIPTION
**Pull request by solberenson**
A `findByOwner` implementation that uses batched calls to
`getMultipleAccountsInfo` (via `Account.getInfos`) instead of
individual calls to `getProgramAccounts` per address.

Notes:
- `amount?.toNumber()` has been replaced with a direct comparison using BN.js operations as `BN.toNumber` is limited to 53 bits and some instances where it overflowed were encountered.
- the return type is `MetadataData[]` instead of `Metadata[]` like the existing `findByOwner` functions

### CODE REVIEW

File: src/programs/metadata/accounts/Metadata.ts [hunk 1]
```diff
@@ -216,6 +217,26 @@ export class Metadata extends Account<MetadataData> {
...
+    const accountsWithAmount = accounts
+      .map(({ data }) => data)
+      // Only include tokens where amount equal to 1.
+      // Note: This is not the same as mint supply.
+      // NFTs by definition have supply of 1, but an account balance > 1 implies a mint supply > 1.
+      .filter(({ amount }) => amount?.eq(new BN(1)));
+
+    const metadataAddresses = await Promise.all(
+      accountsWithAmount.map(async ({ mint }) => await Metadata.getPDA(mint)),
+    );
```
Comment (kespinola) [inline src/programs/metadata/accounts/Metadata.ts:234]:
```suggestion
    const metadataPDALookups = accounts
      // Only include tokens where amount equal to 1.
      // Note: This is not the same as mint supply.
      // NFTs by definition have supply of 1, but an account balance > 1 implies a mint supply > 1.
      .reduce<Promise<T>>((memo, { data }) => {
        if (data.ammount?eq(new BN(1))) {
          return memo;
        }
        
        cont lookup = Metadata.getPDA(data.mint);
        
        return [... memo, lookup];
      }, []);
      
      const metdataAddress = await Promise.all(metadataPDALookup)
```

Possibility to get to 1 loop with doing a reduce.

**Review by kespinola — COMMENTED** [review]
Excited or this one. Currently, Metaplex storefronts lean on this account provider which subscripts to all the token accounts owned by the logged-in user. This opens up a lot of connections to the WS endpoint of the RPC server which isn't well optimized for heavy use of subscriptions.

https://github.com/metaplex-foundation/metaplex/blob/master/js/packages/common/src/contexts/accounts/accounts.tsx#L123

File: src/programs/metadata/accounts/Metadata.ts (hunk 1 above)
Comment (solberenson) [reply src/programs/metadata/accounts/Metadata.ts]: Thanks, I've implemented your suggested change but reversed the logic for the `if` statement as I believe the correct behaviour is to include when the amount is 1.

**Review by m-sebastiyan — APPROVED** [review]

**Review by b2kdaman — APPROVED** [review]

**Review by kespinola — APPROVED** [review]

**Review by kurpav — APPROVED** [review]

**Comment by solberenson** [comment]
> Hmm, are there plans to improve this? This call does `Promise.all` without any limiting — that means for a large wallet this could send 100+ network requests at once which could take down the client / server.
> 
> I've used [p-limit](https://yarnpkg.com/?q=p-limit&p=1) to get around this.
> 
> But an even nicer solution would be to use something like Solana JSON RPC's `getMultipleAccounts` and then to be smart about parsing the data on each account.

Apologies for the delayed response, I haven't looked at this since it was merged.

This solution already uses `getMultipleAccountsInfo` in batches of 99-100 via `Account.getInfos`.

It makes:
- a single call to `getTokenAccountsByOwner`
- one call to `getMultipleAccountsInfo` per 100 metadata accounts

The `Promise.all` is for deriving the metadata addresses; these do not result in network calls.

**Pull request merged**